### PR TITLE
provide a construction functor using individual functors

### PR DESCRIPTION
**Pull request by mantepse**
We enable sage to find common parents also when symmetric functions are involved. In particular, we provide a functorial construction that takes a commutative (coefficient) ring and produces a (commutative) ring of symmetric functions over this ring.

For Macdonald polynomials and other symmetric functions that involve parameters we cheat a bit: the categories should actually be commutative rings with distinguished elements (namely, the parameters). However, having these is very likely overkill.

This is an alternative to #37220, relying on individual construction functors instead of passing around strings.

### CODE REVIEW

File: src/sage/combinat/sf/dual.py [hunk 1]
```diff
@@ -26,7 +26,16 @@
 
 
 class SymmetricFunctionAlgebra_dual(classical.SymmetricFunctionAlgebra_classical):
-    def __init__(self, dual_basis, scalar, scalar_name="", basis_name=None, prefix=None):
+    @staticmethod
+    def __classcall__(cls, dual_basis, scalar, scalar_name="", basis_name=None, prefix=None):
```
Comment (tscrim) [inline src/sage/combinat/sf/dual.py:30]: This needs a doctest (e.g., checking the uniqueness).

File: src/sage/combinat/sf/dual.py [hunk 2]
```diff
@@ -890,6 +927,72 @@ def expand(self, n, alphabet='x'):
...
+    A constructor for algebras of symmetric functions constructed by
+    duality.
+
+    EXAMPLES::
+
+        sage: w = SymmetricFunctions(ZZ).witt()
+        sage: w.dual_basis().construction()
+        (SymmetricFunctionsFunctor[dual Witt], Integer Ring)
+    """
+    def __init__(self, basis):
```
Comment (tscrim) [inline src/sage/combinat/sf/dual.py:942]: This needs a doctest (as always, a `TestSuite(F).run()` is good).

File: src/sage/combinat/sf/hall_littlewood.py [hunk 3]
```diff
@@ -77,7 +77,14 @@ def __repr__(self):
         """
         return self._name + " over %s" % self._sym.base_ring()
 
-    def __init__(self, Sym, t='t'):
+    @staticmethod
+    def __classcall__(cls, Sym, t='t'):
```
Comment (tscrim) [inline src/sage/combinat/sf/hall_littlewood.py:81]: This needs a doctest (such as testing uniqueness).

File: src/sage/combinat/sf/hall_littlewood.py [hunk 4]
```diff
@@ -387,6 +394,24 @@ def __init__(self, hall_littlewood):
...
+        that ``F(R)`` returns ``self``.
+
+        EXAMPLES::
+
+            sage: P = SymmetricFunctions(QQ).hall_littlewood(t=2).P()
+            sage: P.construction()
+            (SymmetricFunctionsFunctor[Hall-Littlewood P with t=2], Rational Field)
+        """
+
+        return (HallLittlewoodSymmetricFunctionsFunctor(self,
```
Comment (tscrim) [inline src/sage/combinat/sf/hall_littlewood.py:410]:
```suggestion
        """
        return (HallLittlewoodSymmetricFunctionsFunctor(self,
```

File: src/sage/combinat/sf/hall_littlewood.py [hunk 5]
```diff
@@ -673,6 +698,80 @@ def scalar_hl(self, x, t=None):
                                                        orthogonal=True)
 
 
+from sage.combinat.sf.sfa import SymmetricFunctionsFunctor
+class HallLittlewoodSymmetricFunctionsFunctor(SymmetricFunctionsFunctor):
+    def __init__(self, basis, name, t):
+        r"""Initialise the functor.
```
Comment (tscrim) [inline src/sage/combinat/sf/hall_littlewood.py:704]:
```suggestion
        r"""
        Initialize the functor.
```

File: src/sage/combinat/sf/hecke.py [hunk 6]
```diff
@@ -121,8 +121,14 @@ class HeckeCharacter(SymmetricFunctionAlgebra_multiplicative):
     - [Ram1991]_
     - [RR1997]_
     """
+    @staticmethod
+    def __classcall__(cls, Sym, q='q'):
```
Comment (tscrim) [inline src/sage/combinat/sf/hecke.py:125]: This needs a doctest (test for the uniqueness).

File: src/sage/combinat/sf/jack.py [hunk 7]
```diff
@@ -50,8 +50,14 @@
 
 
 class Jack(UniqueRepresentation):
+    @staticmethod
```
Comment (tscrim) [inline src/sage/combinat/sf/jack.py:53]: This needs a doctest.

File: src/sage/combinat/sf/llt.py [hunk 8]
```diff
@@ -91,8 +91,14 @@ class LLT_class(UniqueRepresentation):
         sage: HS3x(HC3t2[3,1])
         2*HSp3[3, 1] + (-2*x+1)*HSp3[4]
     """
+    @staticmethod
+    def __classcall__(cls, Sym, k, t='t'):
```
Comment (tscrim) [inline src/sage/combinat/sf/llt.py:95]: This needs a doctest.

File: src/sage/combinat/sf/macdonald.py [hunk 9]
```diff
@@ -97,7 +97,14 @@ def __repr__(self):
         """
         return self._name
 
-    def __init__(self, Sym, q='q', t='t'):
+    @staticmethod
+    def __classcall__(cls, Sym, q='q', t='t'):
```
Comment (tscrim) [inline src/sage/combinat/sf/macdonald.py:101]: This needs a doctest.

File: src/sage/combinat/sf/orthotriang.py [hunk 10]
```diff
@@ -49,8 +49,14 @@ class SymmetricFunctionAlgebra_orthotriang(sfa.SymmetricFunctionAlgebra_generic)
 
     class Element(sfa.SymmetricFunctionAlgebra_generic.Element):
         pass
+    @staticmethod
+    def __classcall__(cls, Sym, base, scalar, prefix, basis_name, leading_coeff=None):
```
Comment (tscrim) [inline src/sage/combinat/sf/orthotriang.py:53]: This doesn't serve any purpose, so you can revert this change to this file.

File: src/sage/combinat/sf/sf.py [hunk 11]
```diff
@@ -976,8 +975,6 @@ def witt(self, coerce_h=True, coerce_e=False, coerce_p=False):
         from . import witt
         return witt.SymmetricFunctionAlgebra_witt(self, coerce_h=coerce_h, coerce_e=coerce_e, coerce_p=coerce_p)
     w = witt
-    # Currently needed by sfa.GradedSymmetricFunctionsBases.corresponding_basis_over
-    Witt = witt
```
Comment (tscrim) [inline src/sage/combinat/sf/sf.py:980]: This is breaking backwards compatibility and creating a slightly inconsistency.

File: src/sage/combinat/sf/witt.py [hunk 12]
```diff
@@ -396,8 +396,14 @@ class SymmetricFunctionAlgebra_witt(multiplicative.SymmetricFunctionAlgebra_mult
         ....:      #this holds for all odd i and is easily proven by induction
         True
     """
+    @staticmethod
+    def __classcall__(cls, Sym, coerce_h=True, coerce_e=False, coerce_p=False):
```
Comment (tscrim) [inline src/sage/combinat/sf/witt.py:400]: This change can be reverted here.

File: src/sage/combinat/sf/character.py [hunk 13]
```diff
@@ -151,6 +168,61 @@ def _b_power_k(self, k):
...
+            parameterized functors are currently not available.
+
+        EXAMPLES::
+
+            sage: from sage.combinat.sf.character import CharacterSymmetricFunctionsFunctor
+            sage: B = SymmetricFunctions(QQ).ht()
+            sage: CharacterSymmetricFunctionsFunctor(B, B.basis_name())
+            SymmetricFunctionsFunctor[induced trivial symmetric group character]
+
+        """
```
Comment (tscrim) [inline src/sage/combinat/sf/character.py:195]:
```suggestion
            SymmetricFunctionsFunctor[induced trivial symmetric group character]
        """
```

File: src/sage/combinat/sf/character.py [hunk 14]
```diff
@@ -151,6 +168,61 @@ def _b_power_k(self, k):
                                     for d in divisors(k))
 
 
+from sage.combinat.sf.sfa import SymmetricFunctionsFunctor
+class CharacterSymmetricFunctionsFunctor(SymmetricFunctionsFunctor):
+    def __init__(self, basis, name):
```
Comment (tscrim) [inline src/sage/combinat/sf/character.py:173]: I would move the `__init__` docstring to the class level.

File: src/sage/combinat/sf/character.py [hunk 15]
```diff
@@ -37,6 +37,23 @@
 
 
 class generic_character(SFA_generic):
+
+    def construction(self):
```
Comment (tscrim) [inline src/sage/combinat/sf/character.py:41]:
```suggestion
class generic_character(SFA_generic):
    def construction(self):
```

File: src/sage/combinat/sf/character.py [hunk 16]
```diff
@@ -151,6 +168,61 @@ def _b_power_k(self, k):
                                     for d in divisors(k))
 
 
+from sage.combinat.sf.sfa import SymmetricFunctionsFunctor
+class CharacterSymmetricFunctionsFunctor(SymmetricFunctionsFunctor):
```
Comment (tscrim) [inline src/sage/combinat/sf/character.py:172]: Every basis should come with a `_prefix` (since they are all subclasses of CFM). So I don't think this functor is necessary.

**Review by tscrim — CHANGES_REQUESTED** [review]
There is one annoying thing: when we override `__eq__` we also need to override `__hash__` to keep things hashable. Perhaps we should take a middle ground to your other proposal. My proposal is we have `SymmetricFunctionsFunctor` that is an endofunctor on `CommutativeRings` with comparison of bases, but then a subclass `SymmetricFunctionsFixedElementsFunctor` (better name likely necessary) that also takes a fixed number of extra arguments for the comparison (and hash). This would then be the base class for, e.g., `MacdonaldSymmetricFunctionsFunctor` and reduce the number of `__eq__` and `__hash__` methods (maybe also `__init__`, but I do like the explicitness), and `.. WARNING::`s necessary. It also fits more mathematically what is going on. What do you think?

Also, we use American English spelling, so `initialise` -> `initialize`.

**Comment by mantepse** [comment]
I am not completely sure whether I did what you had in mind, @tscrim.  In any case, there are no `*args`, as mentioned in https://github.com/sagemath/sage/pull/37220#issuecomment-2021288584, anymore.

To answer the question
> Alternatively, why have the, e.g., `DualBasisFunctor`?

In https://github.com/sagemath/sage/compare/develop...mantepse:sage:sf/construction-functional there are four functors:

* `SymmetricFunctionsFunctor` for bases that are created as `SymmetricFunctions(R).basis(*args)`
* `SymmetricFunctionsFamilyFunctor` for bases that are created as `SymmetricFunctions(R).family(*args).basis()`
* `DualBasisFunctor` for bases that are created as the dual of a basis
* `OrthotriangBasisFunctor`, for bases that are created by orthogonality and triangularity.

It seems to me that these four are quite distinct.

**Comment by tscrim** [comment]
This is what I was thinking.

It seems to me like the last two are basically just the first one that can be passed from `construction` through the arguments.

I am thinking you are right with attaching a list of arguments to the functor with some generality, but I think we should assume it is non-empty.

**Comment by mantepse** [comment]
> This is what I was thinking.
> 
> It seems to me like the last two are basically just the first one that can be passed from `construction` through the arguments.
> 
> I am thinking you are right with attaching a list of arguments to the functor with some generality, but I think we should assume it is non-empty.

Sorry, it's not clear to me what you mean.  I'm in a meeting now for something like half an hour, but maybe we could talk afterwards?

File: src/sage/combinat/sf/character.py (hunk 16 above)
Comment (mantepse) [reply src/sage/combinat/sf/character.py:172]: The strange thing here is that `induced_trivial_character_basis.__init__` takes prefix as a required argument.  Maybe this could or should be made more uniform.  (My previous approach required that the arguments were provided exactly as needed for this `__init__`.)
